File: Python/0200.py
from typing import List

class Solution:
    def dfs(self, posx, posy):        
        if posx == self.rows or posy == self.cols or posx == -1 or \
        posy == -1 or self.grid[posx][posy] == "0": return
        
        self.grid[posx][posy] = "0"

        self.dfs(posx+1, posy)
        self.dfs(posx-1, posy)
        self.dfs(posx, posy+1)
        self.dfs(posx, posy-1)

        return

    def numIslands(self, grid: List[List[str]]) -> int:
        self.grid = grid
        ans = 0
        self.rows = len(grid)
        self.cols = len(grid[0])

        for i in range(self.rows):
            for j in range(self.cols):
                if self.grid[i][j] == "1":
                    ans += 1
                    self.dfs(i, j)

        return ans 

# Topic: Dfs
# Time Complexity: O(n*m)
# Space Complexity: O(1)

File: C++/0200.cpp
#include <vector>

class Solution {
public:
    void dfs(std::vector<std::vector<char>>& grid, int posx, int posy, int rows, int cols){

        if(posx == rows || posy == cols || posx == -1 || posy == -1 || grid[posx][posy] == '0') return;

        grid[posx][posy] = '0';

        dfs(grid, posx+1, posy, rows, cols);
        dfs(grid, posx-1, posy, rows, cols);
        dfs(grid, posx, posy+1, rows, cols);
        dfs(grid, posx, posy-1, rows, cols);

        return;

    }

    int numIslands(std::vector<std::vector<char>>& grid) {
        int ans = 0;
        int rows = grid.size();
        int cols = grid[0].size();

        for(int i = 0; i < rows; i++){
            for(int j = 0; j < cols; j++){
                if(grid[i][j] == '1'){
                    ans++;
                    dfs(grid, i, j, rows, cols);
                }
            }
        }
        return ans;
    }
};

// Topic: Dfs
// Time Complexity: O(n*m)
// Space Complexity: O(n*m)